client/me/purchases/remove-purchase/remove-domain-dialog/index.jsx

```diff
@@ -32,7 +32,7 @@ class RemoveDomainDialog extends Component {
 
 	renderDomainDeletionWarning( productName ) {
 		const { currentRoute, isGravatarDomain, purchase, translate } = this.props;
-		const domain = purchase.domain;
+		const domain = purchase?.domain;
```

Comment:
Below, `domain` is passed to `domainManagementEdit()` and `domainManagementTransferOut()`. If you follow those functions, you'll see that both require their first argument (`siteName`) to be a string and it's used as a string here:

https://github.com/Automattic/wp-calypso/blob/f5224873fd44c8dec991062316405c30e9fc9953/client/my-sites/domains/paths.js#L36

So if `domain` here ends up as `undefined`, then you'd end up with a URL that has `/undefined` at the end and probably will cause further confusion.

So what should we do if `purchase` doesn't exist?  I can think of two possible fixes: 

1. Make sure `domain` is always a string by using `const domain = purchase?.domain ?? '';`, but this probably still doesn't produce a very good UX for the user.
2. Render nothing for the domain deletion warning if the purchase doesn't exist. `if ( ! purchase ) { return null; }`. This is probably a better fallback if for some reason we end up here with invalid data. If you do that before `const domain = purchase.domain;` then you don't even need the optional chaining.

Reply:
> So if `domain` here ends up as `undefined`, then you'd end up with a URL that has `/undefined` at the end and probably will cause further confusion.
> 
> So what should we do if `purchase` doesn't exist?

@sirbrillig I wonder if those issues can be handled in the `connect()` method of this class since that's where `slug` is defined and where the initial dependency on `purchase` would be?

https://github.com/Automattic/wp-calypso/blob/a2477e73fa4e1d6d0f1d16ee1396f204c8784182/client/me/purchases/remove-purchase/remove-domain-dialog/index.jsx#L272-L281

If we do that way, we can just replace the `slug: getSelectedSiteSlug( state )` with `slug: getSiteSlug( ownProps.purchase.siteId )` so we are getting the slug from the purchase, rather than the selected site.

Reply:
> Do you mean this getSiteSlug?

@sirbrillig I feel that I am the one missing something here :D 

What I meant is that it seems the changes introduced in this PR replaces the use of `slug` as a first parameter in `domainManagementEdit` and `domainManagementTransferOut` with `domain` in those lines:

https://github.com/Automattic/wp-calypso/blob/a2477e73fa4e1d6d0f1d16ee1396f204c8784182/client/me/purchases/remove-purchase/remove-domain-dialog/index.jsx#L61-L64

`slug` in the lines above are coming from the component `props` and defined here (line `280`:

https://github.com/Automattic/wp-calypso/blob/a2477e73fa4e1d6d0f1d16ee1396f204c8784182/client/me/purchases/remove-purchase/remove-domain-dialog/index.jsx#L272-L281

So my understanding is that the reasoning behind the changes in this PR, is that `getSiteSelectedSlug( state )` is returning `null` when we don't have a selected site in this context and this is why the fatals were happeing.

I didn't validate that reasoning, but if that's the case, what I'm suggesting is to change `slug: getSiteSelectedSlug( state )` with `slug: getSiteSlug( state, ownProps.purchase.siteId )` here:

https://github.com/Automattic/wp-calypso/blob/a2477e73fa4e1d6d0f1d16ee1396f204c8784182/client/me/purchases/remove-purchase/remove-domain-dialog/index.jsx#L280

Doing that, should make sure that we would have a valid `slug` as long as we have a valid `ownProps.purchase`, regardless whether we have a selected site or not.

Reply:
Ah! I see. Sorry, I had gotten caught up in the details of the crash, but you've pointed out a way to get a more accurate site slug. That does seem like it might fix the issue (which is the most important thing) even if it doesn't prevent other possible crashes in the future. @Aurorum what do you think?

Reply:
Oh, that's a brilliant idea! I've switched to that instead, which does indeed fix the issue. I've also added a guard to check for `purchase` too (is there ever a case where it could/should be empty or missing?).